### ISSUE TITLE
no pong received in 3 seconds

### ISSUE THREAD

**Issue by dengxuesong** [miguelgrinberg/Flask-SocketIO#454]
emitting event "response_verify" to 0f69ec7f4f1143ca81f30970f7180e5d [/xiyouws]
0f69ec7f4f1143ca81f30970f7180e5d: Sending packet MESSAGE data 2/xiyouws,["response_verify",{"status":200,"message":"ok"}]
0f69ec7f4f1143ca81f30970f7180e5d: Received packet MESSAGE data 2/xiyouws,["yeping","118|0|58772650"]
received event "yeping" from 0f69ec7f4f1143ca81f30970f7180e5d [/xiyouws]
emitting event "yepong" to 0f69ec7f4f1143ca81f30970f7180e5d [/xiyouws]
0f69ec7f4f1143ca81f30970f7180e5d: Sending packet MESSAGE data 2/xiyouws,["yepong","118|0|58772650"]
0f69ec7f4f1143ca81f30970f7180e5d: Received packet PING data None
0f69ec7f4f1143ca81f30970f7180e5d: Sending packet PONG data None
0f69ec7f4f1143ca81f30970f7180e5d: Received packet PING data None
0f69ec7f4f1143ca81f30970f7180e5d: Sending packet PONG data None
0f69ec7f4f1143ca81f30970f7180e5d: Received packet MESSAGE data 2/xiyouws,["yeping","118|0|58772650"]
received event "yeping" from 0f69ec7f4f1143ca81f30970f7180e5d [/xiyouws]
emitting event "yepong" to 0f69ec7f4f1143ca81f30970f7180e5d [/xiyouws]
0f69ec7f4f1143ca81f30970f7180e5d: Sending packet MESSAGE data 2/xiyouws,["yepong","118|0|58772650"]
[uwsgi-websocket] "GET /socket.io/?transport=websocket&gid=118&ssid=0&token=0953a5bc3910b079f5bc11b4a1ea7fa2&uid=58772650&sid=0f69ec7f4f1143ca81f30970f7180e5d" (127.0.0.1) no PONG received in 3 seconds !!!

the client emit yeping ,the server reponse yepong,but when this work fine, suddenly the log show [uwsgi-websocket] no pong received in 3 seconds  and server disconnect the client‘s connect。

**Comment by miguelgrinberg** [comment]
That's an uwsgi specific error that occurs when the websocket connection is not healthy. Does everything work if you use gunicorn or the normal `socketio.run()`?

**Comment by stephenrs** [comment]
@miguelgrinberg Can you elaborate on what you mean by "not healthy". I'm seeing the same problem during what appears from the logs to be a normal and healthy exchange of data across the socket connection. Suddenly, I see "no PONG received in 3 seconds", and the client disconnects.

Just prior to that message I see:

```
a209a01afd7f4217b3c10d5170038688: Received packet PING data None
a209a01afd7f4217b3c10d5170038688: Sending packet PONG data None
```
This seems to suggest that the server sent the PONG, but the client didn't receive it (?)

Since I've implemented reconnection logic on the client side, things fix themselves, but this disconnect/reconnect cycle continues to repeat endlessly and consistently.

**Comment by miguelgrinberg** [comment]
@stephenrs there are two levels of ping/pong messages here that make this somewhat confusing. At the protocol level, WebSocket has a ping/pong mechanism, which in this case is handled entirely by uWSGI. The "no PONG received in 3 seconds" message comes from this layer, and has no relation to your application or the Socket.IO protocol. This indicates that the connection between the client and the server is not reliable and the reason I said it was "not healthy".

The Socket.IO protocol has its own ping/pong messages as well, which are unrelated to WebSocket's. These are issued regardless of the transport, so they exist with WebSocket and with long-polling clients. The "Received packet PING" and "Sending packet PONG" messages refer to this level, and are handled by the Socket.IO server (the python-socketio package).

So my guess is that in your case there is a working WebSocket connection, during which the Socket.IO protocol works fine. But at some point the underlying connection breaks, or gets too slow/congested and the WebSocket mechanism that keep the connection alive fail.

**Comment by stephenrs** [comment]
Hi @miguelgrinberg Thanks so much for you quick and thoughtful reply. Since I posted my question, I found this: https://github.com/unbit/uwsgi/pull/1261

Does this sound like it could be the source/solution for the problem? Pardon my lack of understanding of this, but I'm having difficultly getting my head around the number of pieces and possible permutations for getting a smoothly functioning websocket setup in place.
